rust/src/jsonbuilder.rs

```diff
@@ -0,0 +1,539 @@
+        Ok(self)
+    }
+
+    /// Set a key and a string value (from bytes) on an object.
+    pub fn set_string_from_bytes(&mut self, key: &str, val: &[u8]) -> Result<&mut Self, JsonError> {
+        match std::str::from_utf8(val) {
+            Ok(s) => self.set_string(key, s),
+            Err(_) => self.set_string(key, "<failed to convert bytes to string>"),
+        }
+    }
+
+    /// Set a key and an unsigned integer type on an object.
+    pub fn set_uint(&mut self, key: &str, val: u64) -> Result<&mut Self, JsonError> {
+        match self.current_state() {
+            State::ObjectNth => {
+                self.buf.push(',');
+            }
+            State::ObjectFirst => {
+                self.set_state(State::ObjectNth);
+            }
+            _ => {
+                return Err(JsonError::InvalidState);
+            }
+        }
+        self.buf.push('"');
+        self.buf.push_str(key);
+        self.buf.push_str("\":");
+        self.buf.push_str(&val.to_string());
+        Ok(self)
+    }
+
+    /// Encode a string into the buffer, escaping as needed.
+    fn encode_string(&mut self, val: &str) -> Result<(), JsonError> {
+        let mut out = vec![0; (val.len() * 2) + 2];
+        let mut offset = 0;
+        out[offset] = b'"';
+        offset += 1;
+        for c in val.chars() {
+            match c {
+                '"' | '\\' | '/' => {
+                    out[offset] = b'\\';
+                    offset += 1;
+                    out[offset] = c as u8;
+                    offset += 1;
+                }
+                '\n' => {
+                    out[offset] = b'\\';
+                    offset += 1;
+                    out[offset] = b'n';
+                    offset += 1;
+                }
+                '\r' => {
+                    out[offset] = b'\\';
+                    offset += 1;
+                    out[offset] = b'r';
+                    offset += 1;
+                }
+                '\t' => {
+                    out[offset] = b'\\';
+                    offset += 1;
+                    out[offset] = b't';
+                    offset += 1;
+                }
+                // Form feed.
+                '\u{000c}' => {
+                    out[offset] = b'\\';
+                    offset += 1;
+                    out[offset] = b'f';
+                    offset += 1;
+                }
+                // Backspace.
+                '\u{0008}' => {
+                    out[offset] = b'\\';
+                    offset += 1;
+                    out[offset] = b'b';
+                    offset += 1;
+                }
+                _ => {
+                    out[offset] = c as u8;
+                    offset += 1
+                }
+            }
+        }
+        out[offset] = b'"';
+        offset += 1;
+        self.buf.push_str(std::str::from_utf8(&out[0..offset])?);
+        Ok(())
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn scjs_new_object() -> *mut JsonBuilder {
```

Comment:
in line with what we do elsewhere, should this be `rs_js_new_object` instead?

Reply:
That depends.. I want the API to be feel "native" when calling from C, prefixing with "rs" doesn't provide that, and now that Rust is required, and we shouldn't have Rust and C implementations of the same thing, I think its less desirable.

We could also name these SCJSNewObject() or something to fit with the C style and just suppress the Rust warnings this would give us as well.